src/garage/tf/algos/batch_polopt2.py

```diff
@@ -0,0 +1,281 @@
+        logger.log('Optimizing policy...')
+        self.optimize_policy(itr, paths)
+        return paths['average_return']
+
+    def log_diagnostics(self, paths):
+        """Log diagnostic information.
+
+        Args:
+            paths (list[dict]): A list of collected paths.
+
+        """
+        logger.log('Logging diagnostics...')
+        self.policy.log_diagnostics(paths)
+        self.baseline.log_diagnostics(paths)
+
+    def process_samples(self, itr, paths):
+        """Return processed sample data based on the collected paths.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths.
+
+        Returns:
+            dict: Processed sample data, with key
+                * observations  : (numpy.ndarray), shape [B * (T), *obs_dims]
+                * actions       : (numpy.ndarray), shape [B * (T), *act_dims]
+                * rewards       : (numpy.ndarray), shape [B * (T), ]
+                * baselines     : (numpy.ndarray), shape [B * (T), ]
+                * returns       : (numpy.ndarray), shape [B * (T), ]
+                * agent_infos   : (dict), see
+                    OnPolicyVectorizedSampler.obtain_samples()
+                * env_infos     : (dict), see
+                    OnPolicyVectorizedSampler.obtain_samples()
+                * paths         : (list[dict]) The original path with
+                    observation or action flattened
+                * average_return: (numpy.float64)
+                where B = batch size, (T) = variable-length of each trajectory
+
+        """
+        baselines = []
+        returns = []
+
+        max_path_length = self.max_path_length
+
+        if self.flatten_input:
+            paths = [
+                dict(
+                    observations=(self.env_spec.observation_space.flatten_n(
+                        path['observations'])),
+                    actions=(
+                        self.env_spec.action_space.flatten_n(  # noqa: E126
+                            path['actions'])),
+                    rewards=path['rewards'],
+                    env_infos=path['env_infos'],
+                    agent_infos=path['agent_infos']) for path in paths
+            ]
+        else:
+            paths = [
+                dict(
+                    observations=path['observations'],
+                    actions=(
+                        self.env_spec.action_space.flatten_n(  # noqa: E126
+                            path['actions'])),
+                    rewards=path['rewards'],
+                    env_infos=path['env_infos'],
+                    agent_infos=path['agent_infos']) for path in paths
+            ]
+
+        if hasattr(self.baseline, 'predict_n'):
+            all_path_baselines = self.baseline.predict_n(paths)
+        else:
+            all_path_baselines = [
+                self.baseline.predict(path) for path in paths
+            ]
+
+        for idx, path in enumerate(paths):
```

Comment:
Is there a reason this and the for loop below can't be merged?

Reply:
They can be merged, will do